### PR TITLE
Improve video controllers caching and retrieving

### PR DESCRIPTION
**Pull request by rhymes**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description

This PR has been a little bit of a ride, through it I found out we have some inefficiencies in https://dev.to/videos (and likely other pages) where we fetch the same page (made by 24 items) 3 times (once in the server and twice in the frontend).

This PR though addresses another set of things:

- `/api/videos` fetches only the fields it needs to render the JSON, doesn't do a N + 1 on the user, has flexible pagination and sets the correct edge caching, removing server side caching

- `/videos` fetches only the fields needed, removes the SQL queries that were issued in the partial resulting in N + 1 and sets the correct edge caching

I did not address the issues with multiple fetchings in `/videos` as it's outside the scope of this PR and it's probably a much bigger refactoring.

The PR is essentially done but it incorporates #5673 so I'm tagging this as WIP until that one is merged and then rebase and force push this one

### CODE REVIEW

**Comment by rhymes** [comment]
Writing down some of my discoveries, I'll open a separate issues once I get to the bottom of these:

if you notice `insertNext` `insertVideos` and `insertArticles` (though in a way I'm not understanding) in `app/assets/javascripts/initializers/initScrolling.js.erb` check if the item has already been inserted:

https://github.com/thepracticaldev/dev.to/blob/3412af977f26a69cb271095808b43b40db229ef2/app/assets/javascripts/initializers/initScrolling.js.erb#L36-L42

https://github.com/thepracticaldev/dev.to/blob/3412af977f26a69cb271095808b43b40db229ef2/app/assets/javascripts/initializers/initScrolling.js.erb#L130-L136

also the fetchers fetch from page 0 and then page 1 which are the same page

File: app/controllers/api/v0/videos_controller.rb [hunk 1]
```diff
@@ -1,21 +1,23 @@
...
     class VideosController < ApiController
-      caches_action :index,
-                    cache_path: proc { |c| c.params.permit! },
-                    expires_in: 10.minutes
       respond_to :json
 
       before_action :cors_preflight_check
       after_action :cors_set_access_control_headers
 
       def index
```
Comment (mstruve) [inline app/controllers/api/v0/videos_controller.rb]: does this need a before_action for the cache control headers?
```ruby
before_action :set_cache_control_headers, only: %i[index]
```

File: app/controllers/api/v0/videos_controller.rb [hunk 2]
```diff
@@ -1,21 +1,23 @@
...
-        @page = params[:page]
-        @video_articles = Article.published.
-          where.not(video: [nil, ""], video_thumbnail_url: [nil, ""]).
-          where("score > ?", -4).
+        page = params[:page]
+        per_page = (params[:per_page] || 24).to_i
+        num = [per_page, 1000].min
+
+        @video_articles = Article.with_video.
+          includes([:user]).
```
Comment (mstruve) [inline app/controllers/api/v0/videos_controller.rb]: 🎉

File: app/views/videos/index.html.erb [hunk 3]
```diff
@@ -30,7 +30,7 @@
             <span class="video-timestamp"><%= video_article.video_duration_in_minutes %></span>
           </div>
           <p><strong><%= video_article.title %></strong></p>
-          <p class="video-username"><%= User.find(video_article.user_id).name %></p>
```
Comment (mstruve) [inline app/views/videos/index.html.erb]: NICE find!

File: app/controllers/api/v0/videos_controller.rb (hunk 1 above)
Comment (rhymes) [reply app/controllers/api/v0/videos_controller.rb]: great catch!

File: app/controllers/api/v0/videos_controller.rb [hunk 4]
```diff
@@ -1,21 +1,25 @@
...
+      before_action :set_cache_control_headers, only: %i[index]
+
       def index
-        @page = params[:page]
-        @video_articles = Article.published.
-          where.not(video: [nil, ""], video_thumbnail_url: [nil, ""]).
-          where("score > ?", -4).
+        page = params[:page]
+        per_page = (params[:per_page] || 24).to_i
+        num = [per_page, 1000].min
```
Comment (joshpuetz) [inline app/controllers/api/v0/videos_controller.rb]: This is lovely refactoring!

File: app/models/article.rb [hunk 5]
```diff
@@ -138,6 +138,8 @@ class Article < ApplicationRecord
 
   scope :feed, -> { published.select(:id, :published_at, :processed_html, :user_id, :organization_id, :title, :path) }
 
+  scope :with_video, -> { published.where.not(video: [nil, ""], video_thumbnail_url: [nil, ""]).where("score > ?", -4) }
```
Comment (joshpuetz) [inline app/models/article.rb]: 😍

Comment (vaidehijoshi) [reply app/models/article.rb]: 👍

File: app/controllers/api/v0/videos_controller.rb [hunk 6]
```diff
@@ -1,21 +1,25 @@
 module Api
   module V0
     class VideosController < ApiController
-      caches_action :index,
```
Comment (vaidehijoshi) [inline app/controllers/api/v0/videos_controller.rb]: Lol, I had never seen `caches_action` before and had to look up where it was coming from 🙈 

TIL about `actionpack-action_caching` and the fact that we use `caches_action` in `podcast_episodes_controller` and `tags_controller`.  I'm assuming we can move away from that and just use `set_cache_control_headers` in place of it?

Comment (rhymes) [reply app/controllers/api/v0/videos_controller.rb]: @vaidehijoshi this is exactly what i've been doing with these latest PRs :) only 2 endpoints left and we can remove that gem :D

Comment (vaidehijoshi) [reply app/controllers/api/v0/videos_controller.rb]: Rad!! Thanks for doing this 🙌

**Pull request merged**